### ISSUE TITLE
Override inherited choice when restriction effectively removes it

### ISSUE THREAD

**Issue by skinkie** [tefra/xsdata#821]
This issue happens when #820 is merged.

Considering the XSD below. ComplexType R inherits B, but it is restriction it effectivly states that the first choice only valid option becomes x:foo, and the second choice only valid option d1. The code that #820 would introduce Foo because it would not be optional anymore, it does not introduce d1 because the elements are not effectively changed.

It this specific case we would like to override the generated foo_or_c2 and d1_or_d2. I wonder if there is an elegant way to remove the attribute from an inherited dataclass. Or that the only choice would be to completely override the choice by effectively limiting it. 

```xml
<?xml version="1.0"?>
<xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema" targetNamespace="http://xsdtesting" xmlns:x="http://xsdtesting" xmlns:imp="http://importedXSD">
	<xsd:element name="foo" final="#all"/>
	<xsd:complexType name="B">
		<xsd:sequence>
			<xsd:choice minOccurs="1" maxOccurs="1">
				<xsd:element ref="x:foo"/>
				<xsd:element name="c2" minOccurs="1" maxOccurs="1"/>
			</xsd:choice>
			<xsd:choice minOccurs="1" maxOccurs="1">
				<xsd:element name="d1" minOccurs="1" maxOccurs="1" block="#all"/>
				<xsd:element name="d2" minOccurs="1" maxOccurs="1" block="#all"/>
			</xsd:choice>
		</xsd:sequence>
	</xsd:complexType>
	<xsd:complexType name="R">
		<xsd:complexContent>
			<xsd:restriction base="x:B">
				<xsd:sequence>
					<xsd:element ref="x:foo"/>
					<xsd:element name="d1" minOccurs="1" maxOccurs="1" block="#all"/>
				</xsd:sequence>
			</xsd:restriction>
		</xsd:complexContent>
	</xsd:complexType>
	<xsd:element name="doc">
		<xsd:complexType>
			<xsd:choice>
				<xsd:element name="elem" type="x:R"/>
			</xsd:choice>
		</xsd:complexType>
	</xsd:element>
</xsd:schema>
```

```py
@dataclass
class B:
    foo_or_c2: Optional[object] = field(
        default=None,
        metadata={
            "type": "Elements",
            "choices": (
                {
                    "name": "foo",
                    "type": Foo,
                    "namespace": "http://xsdtesting",
                },
                {
                    "name": "c2",
                    "type": object,
                    "namespace": "",
                },
            ),
        }
    )
    d1_or_d2: Optional[object] = field(
        default=None,
        metadata={
            "type": "Elements",
            "choices": (
                {
                    "name": "d1",
                    "type": object,
                    "namespace": "",
                },
                {
                    "name": "d2",
                    "type": object,
                    "namespace": "",
                },
            ),
        }
    )


@dataclass
class R(B):
    foo: Optional[Foo] = field(
        default=None,
        metadata={
            "type": "Element",
            "namespace": "http://xsdtesting",
            "required": True,
        }
    )
```

Think this is what is likely the 'easiest' way. Ideally it wouldn't be a choice, and would reference foo and d1 directly _without_ the choices annotation. It might (and likely is) a bit strange if the attribute name would then stick as foo_or_c2 and d1_or_d2.
```py
@dataclass
class R(B):
    foo_or_c2: Optional[object] = field(
        default=None,
        metadata={
            "type": "Elements",
            "choices": (
                {
                    "name": "foo",
                    "type": Foo,
                    "namespace": "http://xsdtesting",
                }
            ),
        }
    )
    d1_or_d2: Optional[object] = field(
        default=None,
        metadata={
            "type": "Elements",
            "choices": (
                {
                    "name": "d1",
                    "type": object,
                    "namespace": "",
                }
            ),
        }
    )
```

By a slight modification I am able to render this, but I fail to understand yet why d1 does not get the required: True attribute.
```python
@dataclass
class R(B):
    foo: Optional[Foo] = field(
        default=None,
        metadata={
            "type": "Element",
            "namespace": "http://xsdtesting",
            "required": True,
        }
    )
    d1: Optional[object] = field(
        default=None,
        metadata={
            "type": "Element",
            "namespace": "",
        }
    )
```

**Comment by skinkie** [comment]
I think if we can find an effective way to assign the parent choice, and allow choices to be having a count of 1, it is possible to generate it. But the issue remains, we would need to bring in the name of the base choice in order to override it.

```python
    @classmethod
    def validate_override(cls, target: Class, attr: Attr, source_attr: Attr):
        if source_attr.is_any_type and not attr.is_any_type:
            return

        if attr.is_list and not source_attr.is_list:
            # Hack much??? idk but Optional[str] can't override List[str]
            source_attr.restrictions.max_occurs = sys.maxsize

        if (
            attr.default == source_attr.default
            and bool_eq(attr.fixed, source_attr.fixed)
            and bool_eq(attr.mixed, source_attr.mixed)
            and bool_eq(attr.restrictions.tokens, source_attr.restrictions.tokens)
            and bool_eq(attr.restrictions.nillable, source_attr.restrictions.nillable)
            and bool_eq(attr.restrictions.is_optional, source_attr.restrictions.is_optional)
            and bool_eq(attr.restrictions.is_prohibited, source_attr.restrictions.is_prohibited)
            and bool_eq(attr.restrictions.choice, source_attr.restrictions.choice)
        ):
            cls.remove_attribute(target, attr)

        if not bool_eq(attr.restrictions.choice, source_attr.restrictions.choice):
            attr.restrictions.choice = source_attr.restrictions.choice
            attr.name = 'rename'
```

```
@dataclass
class R(B):
    foo: Optional[object] = field(
        default=None,
        metadata={
            "type": "Elements",
            "choices": (
                {
                    "name": "foo",
                    "type": Foo,
                    "namespace": "http://xsdtesting",
                },
            ),
            "required": True,
        }
    )
    d1: Optional[object] = field(
        default=None,
        metadata={
            "type": "Elements",
            "choices": (
                {
                    "name": "d1",
                    "type": object,
                    "namespace": "",
                },
            ),
        }
    )
```

**Comment by tefra** [comment]
Fixed on #820 and #828
